### PR TITLE
Invoke generators in the CompilationTracker

### PR DESCRIPTION
**Pull request by jasonmalinowski**
This moves the invocation of generators into the CompilationTracker so generators are properly included in all Compilations handed out by the Workspace. This will allows us to take advantage of the incremental update support of the compiler APIs.

The TrackedGeneratorDriver is a wrapper type to track some information that right now is not tracked by GeneratorDriver for now; cleaning up that API will be addressed in a future PR and the TrackedGeneratorDriver type will go away.

### CODE REVIEW

**Comment by jasonmalinowski** [comment]
@chsienki This is my current prototype to get the invocation of the generators moved into the correct layer so we have better caching/generation support. This isn't quite done yet as I'm not actually trying to call TryApplyEdits yet, as how the CompilationTracker works a bit more funky than I expected. I'm about 50% sure this is better than your prototype, but it's still not remotely "good" yet. (It's also possible it's worse... :smile:)

**Comment by jasonmalinowski** [comment]
@dotnet/roslyn-ide: some background on this: when @chsienki started doing the source generator work, he did a quick hack to get source generators to run after we produced a compilation in the compilation tracker. Since this was happening _after_ the core workspace processed stuff, this had various problems: TryGetCompilation and GetCompilationAsync would have returned different things, P2P references probably would have been broken, etc. This tries to move the core running into the compilation tracker so there aren't so many things broken with it.

This is still a prototype although one that compiler is interested in dogfooding. The primary goals are:

1. The compilation tracker should behave the same as usual if you aren't on the preview langver, since this all gets short-circuited there.
2. Starting to test out the APIs so we can get some initial experimentation with them.

Out of scope of this PR is:

1. Fleshing out the prototype (you'll see a number of places where the compiler hasn't implemented something, and so we have to punt work because of that.)
2. Trying to use the compiler's incremental update API -- that'll take more work.
3. Any sort of remotely smart performance handling.
4. Any representation of these in a smart way in the Workspaces layer.

Once this is in to unblock compiler dogfooding, then we'll tackle those. This is emphatically a "look, we're walking now" PR where the goal is just to make things be Not Entirely Wrong(tm).

@chsienki and I have been going back and forth on this so we now have some additional feedback on his APIs. Expect those PROTOTYPE comments to go away.

**Comment by jasonmalinowski** [comment]
And oh there was one more goal: re-learn (at least for me) how the CompilationTracker even works, since any source generator design is potentially going to have non-trivial impacts to it. This I think is the "smallest" impact to it in a way that is functionally correct while trying to not be smart about perf in any way. But now that I've paged back into memory how it works I can have some hope at estimating tasks for actually making things smarter too.

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/ISolutionExtensions.cs [hunk 1]
```diff
@@ -56,5 +56,16 @@ public static Document GetRequiredDocument(this Solution solution, DocumentId do
 
             return document;
         }
+
+        public static Document GetRequiredDocument(this Project project, DocumentId documentId)
+        {
+            var document = project.GetDocument(documentId);
+            if (document == null)
```
Comment (jaredpar) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/ISolutionExtensions.cs]: Embrace the `is null` lifestyle :smile:

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs [hunk 2]
```diff
@@ -516,7 +528,7 @@ private async Task<Compilation> GetOrBuildDeclarationCompilationAsync(SolutionSt
                     compilation = compilation.AddSyntaxTrees(trees);
                     trees.Free();
 
-                    WriteState(new FullDeclarationState(compilation), solution);
+                    WriteState(new FullDeclarationState(compilation, new TrackedGeneratorDriver(generatorDriver: null)), solutionServices);
```
Comment (jaredpar) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs]: Nit: rest of PR uses `new TrackedGenerator(null)`. Not opposed to the name but seems like it should be consistent. 

As an aside the pattern seems odd. Feels like it would be better as a named factory of sorts say `TrackedGeneratorDriver.CreateNeedsFullGeneration`

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs [hunk 3]
```diff
@@ -653,9 +669,39 @@ public CompilationInfo(Compilation compilation, bool hasSuccessfullyLoaded)
...
+                    {
+                        var generators = this.ProjectState.AnalyzerReferences.SelectMany(a => a.GetGenerators()).ToImmutableArray();
+                        var additionalTexts = this.ProjectState.AdditionalDocumentStates.Values.SelectAsArray(a => (AdditionalText)new AdditionalTextWithState(a));
+                        var compilationFactory = this.ProjectState.LanguageServices.GetRequiredService<ICompilationFactoryService>();
+
+                        generatorDriver = new TrackedGeneratorDriver(
+                            compilationFactory.CreateGeneratorDriver(
+                                this.ProjectState.ParseOptions!,
+                                generators,
+                                additionalTexts));
```
Comment (jaredpar) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs]: This will drop the value of `generatorDriver.NeedsFullGeneration` if it was `false`. Is there ever a case where that could happen?

Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs]: The NeedsFullGeneration flag is to track whether or not the GeneratorDriver.TryApplyEdits would be useful; that is only useful if you had an existing generator around, so once we're creating a new one we need a full generation no matter what.

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/ISolutionExtensions.cs (hunk 1 above)
Comment (jasonmalinowski) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/ISolutionExtensions.cs]: image

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs (hunk 2 above)
Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs]: TrackedGeneratorDriver is a workaround for a API problem right now with GeneratorDriver. GeneratorDriver has a way to reuse a previous driver with a known set of edits, which you can then apply with TryApplyEdits. But if there's some change that happens to a compilation that doesn't have a corresponding edit, it's just up to the user of the driver to remember to not call TryApplyEdits. I've asked @chsienki to instead offer some API to "poison" a generator driver which would mean:

1. Any stored edits are tossed, because there's no point in holding onto them anymore. (It's otherwise potentially just leaking memory.)
2. Any future calls to TryApplyEdits just fails, so a full generation has to be ran instead. Since the driver requires that as a fallback no matter what (the host has to be ready for TryApplyEdits to fail at any time, so this isn't extra work for us.)

At this point the struct is just going away.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.TrackedGeneratorDriver.cs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+using System.Collections.Immutable;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal partial class SolutionState
+    {
+        // PROTOTYPE: this mostly exists as a way to track if a GeneratorDriver has been mutated in a way
+        // which TryApplyEdit can no longer be used. We should just revise GeneratorDriver to better handle
+        // this pattern.
+        public readonly struct TrackedGeneratorDriver
```
Comment (chsienki) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.TrackedGeneratorDriver.cs]: I think I understand the intent of this class, but it doesn't actually seem to be used? If we're not checking the NeedsFullGeneration property, why is it needed?

Comment (chsienki) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.TrackedGeneratorDriver.cs]: Ah I just saw your comments in response to Jared. The 'posion' api is trivial. Do we want to add it into this PR and drop the tracker here instead?

Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.TrackedGeneratorDriver.cs]: Yes, I tried to use it and then realized I have further problems to sort out. 😦 So it's left in so we *can* use it, just not quite yet.

Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.TrackedGeneratorDriver.cs]: (The problems are mostly that I need further understand how exactly the compilation tracker reuses compilations from full states, versus proper in-progress states. It's a bit funky.)

Comment (chsienki) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.TrackedGeneratorDriver.cs]: Ah ok, makes sense

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/ISolutionExtensions.cs (hunk 1 above)
Comment (jaredpar) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/ISolutionExtensions.cs]: image

File: src/Workspaces/Core/Portable/Workspace/Host/CompilationFactory/ICompilationFactoryService.cs [hunk 5]
```diff
@@ -10,12 +12,9 @@ namespace Microsoft.CodeAnalysis.Host
     internal interface ICompilationFactoryService : ILanguageService
     {
         Compilation CreateCompilation(string assemblyName, CompilationOptions options);
-        Compilation CreateSubmissionCompilation(string assemblyName, CompilationOptions options, Type hostObjectType);
-        Compilation GetCompilationFromCompilationReference(MetadataReference reference);
```
Comment (genlu) [inline src/Workspaces/Core/Portable/Workspace/Host/CompilationFactory/ICompilationFactoryService.cs]: Why are these removed? Is it related to this PR?

File: src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs [hunk 6]
```diff
@@ -592,7 +592,19 @@ public Solution WithProjectMetadataReferences(ProjectId projectId, IEnumerable<M
         /// specified analyzer reference.
         /// </summary>
         public Solution AddAnalyzerReference(ProjectId projectId, AnalyzerReference analyzerReference)
-            => AddAnalyzerReferences(projectId, ImmutableArray.Create(analyzerReference));
+        {
+            if (projectId == null)
```
Comment (genlu) [inline src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs]: Isn't nullalbe enabled here?

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs [hunk 7]
```diff
@@ -653,9 +675,44 @@ public CompilationInfo(Compilation compilation, bool hasSuccessfullyLoaded)
...
+                    var compilationWithoutGeneratedFiles = compilation;
+
+                    if (generatorDriver.GeneratorDriver != null)
+                    {
+                        generatorDriver.GeneratorDriver.RunFullGeneration(compilation, out compilation, cancellationToken);
+                    }
+
+                    RecordAssemblySymbols(compilation, metadataReferenceToProjectId);
+
+                    this.WriteState(
```
Comment (genlu) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs]: OK, just trying to understand this change, we basically made the after-generation compilation the only possible final compilation of state to make everything consistent?

File: src/Workspaces/Core/Portable/Workspace/Host/CompilationFactory/ICompilationFactoryService.cs (hunk 5 above)
Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Workspace/Host/CompilationFactory/ICompilationFactoryService.cs]: Turns out these were unused; it was easier to remove them than null annotate them.

File: src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs (hunk 6 above)
Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs]: Yes, but this is public API surface area.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs (hunk 7 above)
Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs]: The CompilationTracker always moved Compilations through a bunch of states, finally reaching the "Final State". Any call to GetCompilationAsync() would simply push through to the FinalState and then use that one. The FinalState was also the only one that had P2P references wired up as you can see, so essentially that's strengthened here a bit.

The main change was that any given state is also allowed to provide a "good enough" compilation that it has at any point. That could be some partial compilation, or in the case of a FinalState it _is_ the final compilation. I'm now being more explicit and FinalState (if there are generators) now holds onto both the final compilation (post generator) and also the pre-generator compilation, where the latter is used for another later generation pass.

We'll probably have to tweak this further because the generator driver is also going to allow us to have incremental updating, which means we'll need to pull a post-generation compilation through some transformation states as well.

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/ISolutionExtensions.cs (hunk 1 above)
Comment (tmat) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/ISolutionExtensions.cs]: Let me fix that for you:

`=> project.GetDocument(documentId) ?? throw new InvalidOperationException(WorkspaceExtensionsResources.The_solution_does_not_contain_the_specified_document)`

---
In reply to: [393909656](https://github.com/dotnet/roslyn/pull/42373#discussion_r393909656) [](ancestors = 393909656)

**Comment by tmat** [comment]
@chsienki Why is GeneratorDriverState separate from GeneratorDriver?

**Comment by chsienki** [comment]
Mainly so that `CSharpGeneratorDriver` doesn't have to have copy of all the ctor params. (we've talked about moving everything down into the C# layer though so can combine if/when we do that)

---
In reply to: [604057423](https://github.com/dotnet/roslyn/pull/42373#issuecomment-604057423) [](ancestors = 604057423)

**Comment by tmat** [comment]
Yeah, I think combining these types would be better.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTranslationAction.Actions.cs [hunk 8]
```diff
@@ -124,11 +127,76 @@ public ProjectAssemblyNameAction(string assemblyName)
...
                 {
                     return Task.FromResult(oldCompilation.WithAssemblyName(_assemblyName));
                 }
             }
+
+            internal sealed class AddAnalyzerReferencesAction : CompilationTranslationAction
+            {
+                private readonly IEnumerable<AnalyzerReference> _analyzerReferences;
+
+                public AddAnalyzerReferencesAction(IEnumerable<AnalyzerReference> analyzerReferences)
```
Comment (tmat) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTranslationAction.Actions.cs]: >IEnumerable<AnalyzerReference> analyzerReferences [](start = 51, length = 49)

I'd say these should be ImmutableArrays, but for that you'd need my PR to be merged first :)

File: src/Workspaces/Core/Portable/Workspace/Host/CompilationFactory/ICompilationFactoryService.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Host/CompilationFactory/ICompilationFactoryService.cs]: i'm a fan of pulling those sorts of chnages out into a simultaneous PR that is easy to review/merge.  jsut a thought for the future.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs [hunk 9]
```diff
@@ -182,8 +183,16 @@ public CompilationTracker FreezePartialStateWithTree(SolutionState solution, Doc
                 // The user is asking for an in progress snap.  We don't want to create it and then
                 // have the compilation immediately disappear.  So we force it to stay around with a ConstantValueSource.
                 // As a policy, all partial-state projects are said to have incomplete references, since the state has no guarantees.
-                return new CompilationTracker(inProgressProject,
-                    new FinalState(new ConstantValueSource<Optional<Compilation>>(inProgressCompilation), inProgressCompilation, hasSuccessfullyLoaded: false));
+                // We also will drop our generator driver, since the in-progress state will simply have to be whatever we already had and
+                // we won't run any generators any further.
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs]: so, does tihs mean in-progress will never see any generator files?  would that make in-progress almost always useless?  what about preserving the generated files from the previous state and presuming they're the generated code for the in-progress state?

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs [hunk 10]
```diff
@@ -295,7 +304,13 @@ public CompilationTracker FreezePartialStateWithTree(SolutionState solution, Doc
                 }
 
                 inProgressProject = inProgressProject.AddProjectReferences(newProjectReferences);
-                inProgressCompilation = UpdateCompilationWithNewReferencesAndRecordAssemblySymbols(inProgressCompilation, metadataReferences, metadataReferenceToProjectId);
+
+                if (!Enumerable.SequenceEqual(inProgressCompilation.ExternalReferences, metadataReferences))
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs]: what's an ExtenralReference?

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs [hunk 11]
```diff
@@ -382,8 +397,11 @@ private async Task<Compilation> GetOrBuildDeclarationCompilationAsync(SolutionSt
                             return compilation;
                         }
 
+                        TrackedGeneratorDriver generatorDriver;
+                        (compilation, generatorDriver) = await BuildDeclarationCompilationFromInProgressAsync(solutionServices, (InProgressState)state, compilation, cancellationToken).ConfigureAwait(false);
+
                         // We must have an in progress compilation. Build off of that.
-                        return await BuildDeclarationCompilationFromInProgressAsync(solution, (InProgressState)state, compilation, cancellationToken).ConfigureAwait(false);
+                        return compilation;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs]: odd that the driver is dropped on the floor.. can we doc?

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs [hunk 12]
```diff
@@ -486,8 +504,8 @@ private async Task<Compilation> GetOrBuildDeclarationCompilationAsync(SolutionSt
             {
                 try
                 {
-                    var compilation = await BuildDeclarationCompilationFromScratchAsync(solution, cancellationToken).ConfigureAwait(false);
-                    return await FinalizeCompilationAsync(solution, compilation, cancellationToken).ConfigureAwait(false);
+                    var compilation = await BuildDeclarationCompilationFromScratchAsync(solution.Services, cancellationToken).ConfigureAwait(false);
+                    return await FinalizeCompilationAsync(solution, compilation, new TrackedGeneratorDriver(null), cancellationToken).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs]: doc.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs [hunk 13]
```diff
@@ -549,33 +567,36 @@ private Compilation CreateEmptyCompilation()
...
+                    var inProgressGeneratorDriver = state.GeneratorDriver;
 
                     while (intermediateProjects.Length > 0)
                     {
                         cancellationToken.ThrowIfCancellationRequested();
 
-                        inProgressCompilation = await intermediateProjects[0].action.InvokeAsync(inProgressCompilation, cancellationToken).ConfigureAwait(false);
+                        var compilationTranslationAction = intermediateProjects[0].action;
+                        inProgressCompilation = await compilationTranslationAction.TransformCompilationAsync(inProgressCompilation, cancellationToken).ConfigureAwait(false);
+                        inProgressGeneratorDriver = compilationTranslationAction.TransformGeneratorDriver(inProgressGeneratorDriver);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs]: interesting that this is cancellable.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs [hunk 14]
```diff
@@ -653,9 +675,45 @@ public CompilationInfo(Compilation compilation, bool hasSuccessfullyLoaded)
...
+                    }
+
+                    // We will finalize the compilation by adding full contents here.
+                    // PROTOTYPE: allow finalize compilation to incrementally update a prior version
+                    var compilationWithoutGeneratedFiles = compilation;
+
+                    if (generatorDriver.GeneratorDriver != null)
+                    {
+                        // PROTOTYPE: make an API to expose these diagnostics
+                        generatorDriver.GeneratorDriver.RunFullGeneration(compilation, out compilation, out var diagnostics, cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs]: can this not just return a tuple :)

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs]: how are these diagnostics reported?  i.e. this seems like a pri to get working.

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs]: why are these diagnostics *outside* of hte compilation?  i would assume the new compilation instance would just have these internally and `.GetDiagnostics()` on it would return them...

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTranslationAction.Actions.cs (hunk 8 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTranslationAction.Actions.cs]: yes, i would strongly want these not be IEnumerables.  can that not be done in this PR?

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTranslationAction.Actions.cs [hunk 15]
```diff
@@ -124,11 +127,76 @@ public ProjectAssemblyNameAction(string assemblyName)
...
+                {
+                    _analyzerReferences = analyzerReferences;
+                }
+
+                public override TrackedGeneratorDriver TransformGeneratorDriver(TrackedGeneratorDriver generatorDriver)
+                {
+                    var generators = _analyzerReferences.SelectMany(a => a.GetGenerators()).ToImmutableArray();
+                    return new TrackedGeneratorDriver(generatorDriver.GeneratorDriver?.AddGenerators(generators));
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTranslationAction.Actions.cs]: this type is oddly named.  i'm not sure why it's called AddAnalyzerReferencesAction when it's acutaly adding generators to the generator driver.

It also seems weird to have this have to have `?.`   When do we not have a generator driver?  What does this mean in this case, we're just a no-op?  i don't really get it.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTranslationAction.Actions.cs [hunk 16]
```diff
@@ -124,11 +127,76 @@ public ProjectAssemblyNameAction(string assemblyName)
...
+                }
+
+                public override TrackedGeneratorDriver TransformGeneratorDriver(TrackedGeneratorDriver generatorDriver)
+                {
+                    var generators = _analyzerReferences.SelectMany(a => a.GetGenerators()).ToImmutableArray();
+                    return new TrackedGeneratorDriver(generatorDriver.GeneratorDriver?.RemoveGenerators(generators));
+                }
+            }
+
+            internal sealed class AddAdditionalDocumentsAction : CompilationTranslationAction
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTranslationAction.Actions.cs]: base name is a bit off now.  CompilationTranslationAction.. but it translates compilations and generators.  

do we always have to execute both translation steps in lockstep?  if so, might be better ot have this return a pair of the new compilation and new generator driver.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTranslationAction.Actions.cs [hunk 17]
```diff
@@ -124,11 +127,76 @@ public ProjectAssemblyNameAction(string assemblyName)
...
+
+                public RemoveAdditionalDocumentsAction(ImmutableArray<TextDocumentState> additionalDocuments)
+                {
+                    _additionalDocuments = additionalDocuments;
+                }
+
+                public override TrackedGeneratorDriver TransformGeneratorDriver(TrackedGeneratorDriver generatorDriver)
+                {
+                    // PROTOTYPE: right now we have no way to remove additional files so we have to do one from scratch
+                    return new TrackedGeneratorDriver(generatorDriver: null);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTranslationAction.Actions.cs]: not certain i understand hte comment.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTranslationAction.cs [hunk 18]
```diff
@@ -12,7 +14,20 @@ internal partial class SolutionState
     {
         private abstract partial class CompilationTranslationAction
         {
-            public abstract Task<Compilation> InvokeAsync(Compilation oldCompilation, CancellationToken cancellationToken);
+            public virtual Task<Compilation> TransformCompilationAsync(Compilation oldCompilation, CancellationToken cancellationToken)
+            {
+                return Task.FromResult(oldCompilation);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTranslationAction.cs]: seems like a bug waiting to happen :)   but maybe it's obvious from subclasses they need to overide thsi?

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.TrackedGeneratorDriver.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.TrackedGeneratorDriver.cs]: do you guys think you'll still need this in the future?

File: src/Workspaces/CoreTest/SolutionTests/SolutionWithSourceGeneratorTests.cs [hunk 19]
```diff
@@ -0,0 +1,139 @@
...
+            {
+                // PROTOTYPE: context.RegisterForAdditionalFileChanges(UpdateContext);
+            }
+
+            private void AddSourceForAdditionalFile(SourceGeneratorContext context, AdditionalText file) => context.AddSource(GetGeneratedFileName(file.Path), SourceText.From("", Encoding.UTF8));
+
+            private string GetGeneratedFileName(string path) => $"{Path.GetFileNameWithoutExtension(path)}.generated";
+        }
+    }
+}
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CoreTest/SolutionTests/SolutionWithSourceGeneratorTests.cs]: i appreciate test.  one thing that we may need to invest in IMO is a way to do fine-grained compilation tracker testing. i.e to actually validate that we are or aren't doing specific pieces of work at specific times.

File: src/Workspaces/Core/Portable/Workspace/Host/CompilationFactory/ICompilationFactoryService.cs (hunk 5 above)
Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Workspace/Host/CompilationFactory/ICompilationFactoryService.cs]: This was broken down commit-by-commit; my bad for not mentioning that in the original PR message.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTranslationAction.Actions.cs (hunk 8 above)
Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTranslationAction.Actions.cs]: @tmat merged in #42451 so I'll get that over to the base branch so I can use the new helpers added.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs (hunk 9 above)
Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs]: If something has already pulled the compilation forward to the FinalState then the final state should be used and we're good but yeah that's not great. I'm going to need to redo the GeneratorDriver API to be more explicit about the syntax trees which will then let us carry those along accordingly.

Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs]: I will add a test though that fetching the full compilation and then fetching an in-progress does still have generators, and also will manually test we're not seeing really bad behavior in completion...

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs (hunk 10 above)
Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs]: It's the list of references that excludes references that get added via `#r`. Those are computed automatically and added into the Compilation's direct "References" collection, but in this case we're explicitly only mutating the ones that came elsewhere.

(This code just moved, but it's a great question nonetheless.)

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTranslationAction.Actions.cs (hunk 15 above)
Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTranslationAction.Actions.cs]: I was naming the actions for the overall operation that the express (in this case, adding analyzer references) and leaving the implementation detail of what that actually results _in_ internal.

We don't have a generator driver if a previous operation meant we had to throw it out and recompute it (which in the current prototype is quite a lot.) We _could_ have the caller filter this out but then there's no way to null annotate that this is non-null since the type is generally null. I've logged https://github.com/dotnet/roslyn/issues/42815 to track what will ultimately let us get rid of TrackedGeneratorDriver which then means we can make this only get called if we actually have one to update.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTranslationAction.Actions.cs (hunk 16 above)
Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTranslationAction.Actions.cs]: We don't _have_ to do them both in lockstep but both would be holding onto the same data, and potentially running one but not the other means we need to deal with issues where one has been pulled forward but the other hasn't.

I'll rename the base type in any case.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTranslationAction.cs (hunk 18 above)
Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTranslationAction.cs]: That was my guess -- things will be pretty broken otherwise. :-)

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs (hunk 14 above)
Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs]: Filed https://github.com/dotnet/roslyn/issues/42818 to discuss further.

Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs]: Right now they're unreported indeed -- this PR was originally written prior to that being a thing, so we'll have to do that as a mop-up.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTranslationAction.Actions.cs (hunk 17 above)
Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTranslationAction.Actions.cs]: Reworded to be a bit clearer, hopefully?

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTranslationAction.Actions.cs (hunk 16 above)
Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTranslationAction.Actions.cs]: Renamed.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs (hunk 13 above)
Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs]: What do you mean? I guess the idea is if somebody asks for a compilation and they stop then we stop processing until somebody were to ask again.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs (hunk 11 above)
Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs]: Ah right it seems this was only used for skeleton assembly reference generation which is a bit funky. I'm going to file a bug to delete this entire function since I think there's some cleanup we can do here.

Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs]: https://github.com/dotnet/roslyn/issues/42819 filed. This code all looks fishy in general... :-(

File: src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverTests.cs [hunk 20]
```diff
@@ -348,7 +348,8 @@ class C { }
             // now try apply edits
             driver = driver.TryApplyEdits(compilation, out editedCompilation, out succeeded);
             Assert.True(succeeded);
-            Assert.Equal(6, editedCompilation.SyntaxTrees.Count());
+            // PROTOTYPE: we get five, because the previous edit was forgotten in some cases
```
Comment (ryzngard) [inline src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverTests.cs]: Do we need to add anything tracking this? Or is // PROTOTYPE: fine?

Comment (jasonmalinowski) [reply src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverTests.cs]: Oops that slipped back in. Good catch, thanks!

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs (hunk 9 above)
Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs]: Test was added for that.
